File: JavaVersion/AoC2020/src/com/company/Day4.java
package com.company;

import java.sql.Time;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day4 extends DefaultClass{

    public Day4()
    {
        super(4,true);
        ArrayList<String> inp = fr.ReadLine();
        ArrayList<String> Data = new ArrayList();
        StringBuilder temp=new StringBuilder();
        ResetClock();
        for (int i = 0; i < inp.size(); i++) {
            if(inp.get(i).equals(""))
            {
                Data.add(temp.toString());
                temp = new StringBuilder();
            }
            else
            {
                temp.append(inp.get(i)+" ");
            }
        }
        Data.add(temp.toString());
        int sum=0;
        for (int i = 0; i < Data.size(); i++) {
            if(Data.get(i).contains("byr")&&Data.get(i).contains("iyr")&&Data.get(i).contains("eyr")&&Data.get(i).contains("hgt")&&Data.get(i).contains("hcl")
                    &&Data.get(i).contains("ecl")&&Data.get(i).contains("pid"))
                sum++;
        }
        System.out.println("Part1: "+sum);
        TimeClock();
        ResetClock();
        System.out.println("Part2: "+ part2(Data));
        TimeClock();
    }

    private int part2(ArrayList<String> inp)
    {
        int sum=0;
        for (int i = 0; i < inp.size(); i++) {
            Data x = new Data();
            x.InputString(inp.get(i));
            if (x.Valid())
                sum++;
        }
        return sum;
    }

    private class Data
    {
        public String byr="", iyr="", eyr="", hgt="", hcl="", ecl="", pid="", cid="";

        public void InputString(String str)
        {
            String[] sep = str.split(" ");
            for (int i = 0; i < sep.length; i++) {
                if(sep[i].contains("byr"))
                {
                    String[] sep2 = sep[i].split(":");
                    byr = sep2[1];
                }
                else if(sep[i].contains("iyr"))
                {
                    String[] sep2 = sep[i].split(":");
                    iyr = sep2[1];
                }
                else if(sep[i].contains("eyr"))
                {
                    String[] sep2 = sep[i].split(":");
                    eyr = sep2[1];
                }
                else if(sep[i].contains("hgt"))
                {
                    String[] sep2 = sep[i].split(":");
                    hgt = sep2[1];
                }
                else if(sep[i].contains("hcl"))
                {
                    String[] sep2 = sep[i].split(":");
                    hcl = sep2[1];
                }
                else if(sep[i].contains("ecl"))
                {
                    String[] sep2 = sep[i].split(":");
                    ecl = sep2[1];
                }
                else if(sep[i].contains("pid"))
                {
                    String[] sep2 = sep[i].split(":");
                    pid = sep2[1];
                }
            }

        }

        public boolean Valid()
        {
            return ValidBYR()&&ValidECL()&&ValidEYR()&&ValidHCL()&&ValidIRY()&&ValidPID()&&ValidHGT();
        }

        private boolean ValidBYR()
        {
            try {
                return ((Integer.parseInt(byr) >= 1920 && Integer.parseInt(byr) <= 2002));
            }
            catch (Exception e)
            {
                return false;
            }
        }

        private boolean ValidIRY()
        {
            try {
                return (Integer.parseInt(iyr) >= 2010 && Integer.parseInt(iyr) <= 2020);
            }
            catch (Exception e)
            {
                return false;
            }
        }

        private boolean ValidEYR()
        {
            try {
                return (Integer.parseInt(eyr) >= 2020 && Integer.parseInt(eyr) <= 2030);
            }
            catch (Exception e) {
                return false;
            }
        }
        private boolean ValidHGT()
        {
            try {
                int hg = Integer.parseInt(hgt.substring(0, hgt.length() - 2));
                String cm = hgt.substring(hgt.length() - 2, hgt.length());
                if (cm.equals("cm")) {
                    if (hg >= 150 && hg <= 193)
                        return true;
                    else
                        return false;
                } else if (cm.equals("in")) {
                    if (hg >= 59 && hg <= 76)
                        return true;
                    else
                        return false;
                }
                return false;
            }
            catch (Exception e)
            {
                return false;
            }
        }

        private boolean ValidHCL()
        {
            try {
                if (hcl.charAt(0) == '#') {
                    return isHexadecimal(hcl.substring(1));
                }
            }
            catch (Exception e)
            {
                return false;
            }
            return  false;
        }

        private boolean ValidECL()
        {
            if(ecl.equals("amb")||ecl.equals("blu")||ecl.equals("gry")||ecl.equals("brn")||ecl.equals("grn")||ecl.equals("hzl")||ecl.equals("oth"))
                return true;
            else
                return false;
        }

        private boolean ValidPID()
        {
            if(pid.length()==9)
            {
                try
                {
                    Integer.parseInt(pid);
                    return true;
                }
                catch (Exception e)
                {
                    return false;
                }
            }
            return false;
        }
    }

    private static final Pattern HEXADECIMAL_PATTERN = Pattern.compile("\\p{XDigit}+");

    private boolean isHexadecimal(String input) {
        final Matcher matcher = HEXADECIMAL_PATTERN.matcher(input);
        return matcher.matches();
    }

}

File: CPPVersion/AoC2020CPP/AoC2020CPP/Day4.cpp
#include "Day4.h"

Day4::Day4() : DefaultClass(4, true)
{
	std::vector<std::string> inp = FR.ReadLine();
	std::string temp;
	std::vector<std::string> passports;
	for (int i = 0; i < inp.size(); i++)
	{
		if (inp[i] == "")
		{
			temp.resize(temp.size() - 1);
			passports.push_back(temp);
			temp = "";
		}
		else
			temp.append(inp[i]+" ");
	}
	passports.push_back(temp);
	for (int i = 0; i < passports.size(); i++)
	{
		InputString(&passports[i]);
	}
	std::cout << "File read and extract to data type: ";
	TimeClock();
	ResetClock();
	Part1();
	TimeClock();
	ResetClock();
	Part2();
	TimeClock();
}
	
void Day4::InputString(std::string* str)
{
	 FR.ReplaceChar(str, ' ', ':');
	std::vector<std::string> arr = FR.SplitString(*str,':');
	Data.push_back(PassPort());
	for (int i = 0; i < arr.size(); i+=2)
	{
		if (arr[i] == "byr")
			Data.back().byr = arr[i + 1];
		else if (arr[i] == "iyr")
			Data.back().iyr = arr[i + 1];
		else if (arr[i] == "eyr")
			Data.back().eyr = arr[i + 1];
		else if (arr[i] == "hgt")
			Data.back().hgt = arr[i + 1];
		else if (arr[i] == "hcl")
			Data.back().hcl = arr[i + 1];
		else if (arr[i] == "ecl")
			Data.back().ecl = arr[i + 1];
		else if (arr[i] == "pid")
			Data.back().pid = arr[i + 1];
	}
}

bool Day4::IsValid(PassPort* P)
{
	int byr, iyr, eyr, hi;
	std::string hgt = P->hgt, hcl = P->hcl, ecl = P->ecl, pid = P->pid;
	if (P->byr != "")
		byr = std::stoi(P->byr);
	else
		return false;
	if (P->iyr != "")
		iyr = std::stoi(P->iyr);
	else
		return false;
	if (P->eyr != "")
		eyr = std::stoi(P->eyr);
	else
		return false;
	if (P->hgt != ""&&P->hgt.size()>3)
		hi = std::stoi(hgt.substr(0, hgt.size() - 2));
	else
		return false;

	if (byr < 1920 || byr>2002)
		return false;
	if (iyr < 2010 || iyr>2020)
		return false;
	if (eyr < 2020 || eyr >2030)
		return false;
	if (hgt[hgt.size() - 2] == 'c')
	{
		if (hi < 150 || hi>193)
			return false;
	}
	else if (hgt[hgt.size() - 2] == 'i')
	{
		if (hi < 59 || hi>76)
			return false;
	}
	if (hcl[0] != '#')
		return false;
	else
	{
		for (int i = 1; i < hcl.size(); i++)
		{
			if (!(((int)hcl[i] > 47 && (int)hcl[i] < 58) || ((int)hcl[i] > 96 && (int)hcl[i] < 103)))
				return false;
		}
	}

	if (!(ecl == "amb" || ecl == "blu" || ecl == "brn" || ecl == "gry" || ecl == "grn" || ecl == "hzl" || ecl == "oth"))
		return false;

	if (pid.size() != 9)
		return false;	
	for (int i = 0; i < pid.size(); i++)
	{
		if (!std::isdigit(pid[i]))
			return false;
	}

	return true;
}
void Day4::Part1()
{
	int sum = 0;
	for (int i = 0; i <Data.size(); i++)
	{
		PassPort* p = &Data.at(i);
		if (p->byr != "" && p->iyr != "" && p->eyr != "" && p->hgt != "" && p->hcl != "" && p->ecl != "" && p->pid != "")
			sum++;
	}
	std::cout << "Part1: " << sum<< "\n";
}
void Day4::Part2()
{
	int sum = 0;
	for (int i = 0; i < Data.size(); i++)
	{
		if (IsValid(&Data.at(i)))
			sum++;
	}
	std::cout << "Part2: " << sum << "\n";
}
